library/init/category/lift.lean

```diff
@@ -0,0 +1,78 @@
+/-
+Copyright (c) 2016 Gabriel Ebner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gabriel Ebner, Sebastian Ullrich
+-/
+prelude
+import init.function init.coe
+import init.category.monad
+import init.util
+
+universes u v w
+
+/-- A function for lifting a computation from an inner monad to an outer monad.
+    Like https://hackage.haskell.org/package/transformers-0.5.5.0/docs/Control-Monad-Trans-Class.html,
+    but `n` does not have to be a monad transformer. -/
+class has_monad_lift (m : Type u → Type v) (n : Type u → Type w) :=
+(monad_lift {} : ∀ {α}, m α → n α)
+
+/-- The reflexive-transitive closure of `has_monad_lift`. -/
+class has_monad_lift_t (m : Type u → Type v) (n : Type u → Type w) :=
+(monad_lift {} : ∀ {α}, m α → n α)
+
+export has_monad_lift_t (monad_lift)
+
+@[reducible] def has_monad_lift_to_has_coe {m n} [has_monad_lift_t m n] {α} : has_coe (m α) (n α) :=
+⟨monad_lift⟩
+
+instance has_monad_lift_t_trans (m n o) [has_monad_lift n o] [has_monad_lift_t m n] : has_monad_lift_t m o :=
+⟨λ α ma, has_monad_lift.monad_lift (monad_lift ma : n α)⟩
+
+instance has_monad_lift_t_refl (m) : has_monad_lift_t m m :=
+⟨λ α, id⟩
+
+@[simp] lemma monad_lift_refl {m : Type u → Type v} {α} : (monad_lift : m α → m α) = id := rfl
+
+
+/-- A functor in the category of monads. Can be used to lift monad-transforming functions.
+    Based on https://hackage.haskell.org/package/pipes-2.4.0/docs/Control-MFunctor.html,
+    but not restricted to monad transformers. -/
+class monad_functor (m m' : Type u → Type v) (n n' : Type u → Type w) :=
+(monad_map {} {α : Type u} : (∀ {α}, m α → m' α) → n α → n' α)
+
+/-- The reflexive-transitive closure of `monad_functor`. -/
+class monad_functor_t (m m' : Type u → Type v) (n n' : Type u → Type w) :=
+(monad_map {} {α : Type u} : (∀ {α}, m α → m' α) → n α → n' α)
+
+export monad_functor_t (monad_map)
+
+instance monad_functor_t_trans (m m' n n' o o') [monad_functor n n' o o'] [monad_functor_t m m' n n'] :
+  monad_functor_t m m' o o' :=
+⟨λ α f, monad_functor.monad_map (λ α, (monad_map @f : n α → n' α))⟩
+
+instance monad_functor_t_refl (m m') : monad_functor_t m m' m m' :=
+⟨λ α f, f⟩
+
+@[simp] lemma monad_map_refl {m m' : Type u → Type v} (f : ∀ {α}, m α → m' α) {α} : (monad_map @f : m α → m' α) = f := rfl
+
+
+/-- Run a monad stack to completion. -/
+class monad_run (out : out_param $ Type u → Type v) (m : Type u → Type v) :=
+(run {} {α : Type u} : m α → out α)
```

Comment:
What does "completion" mean here?
BTW, the signature of `has_monad_lift.monad_lift` and `monad_run.run` is essentially the same.
The main difference is that `out` is marked as `out_param` here.

Reply:
Extended the doc string. Btw, I added `unrun` because it seemed like it could be useful, but I haven't needed it yet.
There is an alternative design where the stack is not run "to completion" but stopped at some inner monad `n`. This may be useful for stacks based on io, e.g. `run : state_t st (except_t e io) a -> st -> except e (io a)`.